# Calculation Errors

I can expect that the result of the following program will be `1.0`.

In [1]:
i = 0
ans = 0

while i < 10:
    ans = ans + 0.1
    i = i + 1

print(ans)

0.9999999999999999


But I got `0.9999999999999999`. 

This situation occurs due to the way computers store data. It happens using zeros and ones, meaning that the data storage system in a computer is the binary system. To understand the output of the above program, one must first get a basic understanding of how data is recorded in memory.

# Decimal to Binary Conversion

We divide the integer part entirely by 2 and record the remainder of the division.

**Example 1: Converting a Decimal Number to Binary**

Given the decimal number:

$ n = 69.625 $

**Converting the Integer Part**

We convert the integer part ($69$) to binary using successive division by $2$ and noting the remainders:

$
69 \div 2 = 34 \text{ mod } 1 \\
34 \div 2 = 17 \text{ mod } 0 \\
17 \div 2 = 8 \text{ mod } 1 \\
8 \div 2 = 4 \text{ mod } 0 \\
4 \div 2 = 2 \text{ mod } 0 \\
2 \div 2 = 1 \text{ mod } 0 \\
1 \div 2 = 0 \text{ mod } 1
0
$

Reading the remainders from bottom to top, the integer part in binary is: $ 69_{(10)} = 1000101_{(2)} $.

**Converting the Fractional Part**

To convert the fractional part ($0.625$) to binary, we add $\frac{1}{2^i}$ and extract the integer parts:

$i = 1 \\ ans = 0: \frac{1}{2^1} = \frac{1}{2} = 0.5 \to ans + 0.5 < 0.625 \quad \text{take}$

$i = 2 \\ ans = 0.5: \frac{1}{2^2} = \frac{1}{4} = 0.25 \to ans + 0.25 = 0.75 > 0.625 \quad \text{not take}$

$i = 3 \\ ans = 0.5: \frac{1}{2^3} = \frac{1}{8} = 0.125 \to ans + 0.125 = 0.625 \quad \text{take}$

$i = 4 \\ ans = 0.625: \text{stop}$

The fractional part in binary is read from top to bottom as: $ 0.625_{(10)} = 0.101_{(2)} $.

Combining both parts: $ 69.625_{(10)} = 1000101.101_{(2)} $.


**Example 2.**

$n = 14.09375$

**Converting the Integer Part**

$14 \div 2 = 7 \text{ mod } 0 \\
7 \div 2 = 3 \text{ mod } 1 \\
3 \div 2 = 1 \text{ mod } 1 \\
1 \div 2 = 4 \text{ mod } 1 \\
0$

Reading the remainders from bottom to top, the integer part in binary is: $14|_{(10)} = 1110|_{(2)}$.

**Converting the Fractional Part**

$i = 1 \\ ans = 0: \frac{1}{2^1} = \frac{1}{2} = 0.5 \to ans + 0.5 = 0.5 > 0.09375 \quad \text{not take}$

$i = 2 \\ ans = 0: \frac{1}{2^2} = \frac{1}{4} = 0.25 \to ans + 0.25 = 0.25 > 0.09375 \quad \text{not take}$

$i = 3 \\ ans = 0: \frac{1}{2^3} = \frac{1}{8} = 0.125 \to ans + 0.125 = 0.125 > 0.09375 \quad \text{not take}$

$i = 4 \\ ans = 0: \frac{1}{2^4} = \frac{1}{16} = 0.0625 \to ans + 0.0625 = 0.0625 < 0.09375 \quad \text{take}$

$i = 5 \\ ans = 0.0625: \frac{1}{2^5} = \frac{1}{32} = 0.03125 \to ans + 0.03125 = 0.09375 \quad \text{take}$

$i = 6 \\ ans = 0.09375: \text{stop}$

The fractional part in binary is read from top to bottom as: $0.09375|_{(10)} = 0.00011|_{(2)}$.

Combining both parts: $14.09375|_{(10)} = 1110.00011|_{(2)}$.

# Binary to Decimal Conversion

Converting binary numbers to decimal involves expanding the binary number into its positional components, where each digit is multiplied by the corresponding power of $2$. The powers of $2$ are positive for digits to the left of the decimal point, and negative for digits to the right.

**Example 1:**

Given the binary number:

$ n|_{(2)} = 10110.101|_{(2)} $

We convert it to decimal by expanding it using powers of $2$:

$$
n|_{(10)} = 1 \times 2^4 + 0 \times 2^3 + 1 \times 2^2 + 1 \times 2^1 + 0 \times 2^0 + 1 \times 2^{-1} + 0.0 \times 2^{-2} + 1 \times 2^{-3}
$$

Now, calculate each term:

$$
n|_{(10)} = 16 + 0 + 4 + 2 + 0 + 0.5 + 0.0 + 0.125 = 22.625
$$

So, the decimal representation of the binary number $10110.101_{(2)}$ is:

$$
n|_{(10)} = 22.625
$$

**Example 2:**

Given the binary number:

$ n|_{(2)} = 11101.01011|_{(2)} $

We convert it to decimal by expanding it using powers of $2$:

$$
n|_{(10)} = 1 \times 2^4 + 1 \times 2^3 + 1 \times 2^2 + 0 \times 2^1 + 1 \times 2^0 + 0 \times 2^{-1} + 1 \times 2^{-2} + 0 \times 2^{-3} + 1 \times 2^{-4} + 1 \times 2^{-5}
$$

Now, calculate each term:

$$
n|_{(10)} = 16 + 8 + 4 + 0 + 1 + 0.0 + 0.25 + 0.0 + 0.0625 + 0.03125 = 29.34375
$$

So, the decimal representation of the binary number $11101.01011_{(2)}$ is:

$$
n|_{(10)} = 29.34375
$$

# Two's Complement

The **two's complement** representation is a method used to encode both positive and negative integers in binary form. It is widely used in computer systems for representing signed numbers, as it allows for efficient arithmetic operations like addition and subtraction to be performed without distinguishing between positive and negative values.

**General Approach**:

1. **Positive Numbers**: When the number is positive, the two's complement representation is the same as the regular binary representation. The leading bit (most significant bit) is `0` for positive numbers in an n-bit system.
   
2. **Negative Numbers**: When the number is negative, we first find the binary representation of its absolute value, then take the **bitwise negation** (invert all the bits) and finally **add one** to the result. This operation ensures that negative numbers are represented efficiently in binary, allowing arithmetic operations to work uniformly for both positive and negative numbers.

**Conversion Steps for Negative Numbers**:

- **Step 1**: Find the binary representation of the absolute value of the number.

- **Step 2**: Invert all the bits (bitwise NOT).

- **Step 3**: Add $1$ to the result.

This method ensures that the range of numbers that can be represented in an n-bit system extends from $-2^(n-1)$ to $2^(n-1) - 1$.


**Example 1:**

Convert $14$ into an 8-bit two's complement representation.

1. First, convert $14$ into binary:
   - $14 \div 2 = 7 \, \text{ mod } \, 0$
   - $7 \div 2 = 3 \, \text{ mod } \, 1$
   - $3 \div 2 = 1 \, \text{ mod } \, 1$
   - $1 \div 2 = 0 \, \text{ mod } \, 1$
   
   The binary representation of $14$ is $1110$.
   
2. Since $14$ is positive, the two's complement representation is the same as the binary form, with leading zeros to fit into 8 bits.

Answer:
$$
\text{Two's complement of 14 on 8 bits} = 00001110
$$

**Example 2:**

Convert $-25$ into an 8-bit two's complement representation.

1. First, find the binary representation of the absolute value of $25$:
   - $25 \div 2 = 12 \, \text{ mod } \, 1$
   - $12 \div 2 = 6 \, \text{ mod } \, 0$
   - $6 \div 2 = 3 \, \text{ mod } \, 0$
   - $3 \div 2 = 1 \, \text{ mod } \, 1$
   - $1 \div 2 = 0 \, \text{ mod } \, 1$
   
   The binary representation of $25$ is $11001$.
   
2. Since we are using 8 bits, pad it with leading zeros to make it 8 bits: $00011001$.

3. Now, invert all the bits (bitwise negation):
   $$ 
   \neg(00011001) = 11100110
   $$

4. Finally, add $1$ to the result:
   $$ 
   11100110 + 1 = 11100111
   $$

Thus, the two's complement representation of $-25$ on 8 bits is:
$$
\text{Two's complement of -25 on 8 bits} = 11100111
$$

**Summary**:

- **Positive numbers** are represented directly in binary with a leading `0` for an n-bit system.
- **Negative numbers** are represented by taking the binary of the absolute value, inverting the bits, and adding 1. The most significant bit (MSB) is `1` for negative numbers, which indicates the sign of the number in two's complement notation.
- The two's complement system simplifies binary arithmetic, as both addition and subtraction can be performed without needing separate logic for positive and negative numbers.

**Additional Note**:

The two's complement method allows for **efficient subtraction**, as subtracting a number is equivalent to adding its two's complement. Additionally, this method eliminates the need for a separate sign bit and ensures that zero is uniquely represented (with all bits set to `0`).


# Floating-Point Arithmetic (IEEE 754)

IEEE 754 is the standard for floating-point computation. It defines two common formats:
- **32-bit (Single Precision)**
- **64-bit (Double Precision)**

### 32-bit IEEE 754 Format (Single Precision)
The format for a 32-bit floating-point number is as follows:
- 1 bit for the **sign** (S)
- 8 bits for the **exponent** (E)
- 23 bits for the **fraction (mantissa)** (M)

The formula is: $(-1)^S * (1 + M) * 2^{E - 127}$


### 64-bit IEEE 754 Format (Double Precision)
The format for a 64-bit floating-point number is as follows:
- 1 bit for the **sign** (S)
- 11 bits for the **exponent** (E)
- 52 bits for the **fraction (mantissa)** (M)

The formula is: $(-1)^S * (1 + M) * 2^{E - 1023}$

**Example 1**: Convert -12.25 into IEEE 754 32-bit and 64-bit

**32-bit IEEE 754 (Single Precision)**

1. Convert -12.25 to binary:
   - **12** = `1100`
   - **0.25** = `0.01`
   - So, -12.25 = `-1100.01`

2. Normalize the binary:
   - `-1100.01` = `-1.10001 × 2^3`

3. Encode:
   - Sign bit (S) = `1` (negative)
   - Exponent (E) = `3 + 127 = 130` → binary `10000010`
   - Mantissa (M) = `10001000000000000000000` (take the fractional part after the decimal point)

So, the IEEE 754 32-bit representation is:
| Sign | Exponent   | Mantissa                                           |
|------|------------|----------------------------------------------------|
| 1    | 10000010   | 10001000000000000000000                           |



**64-bit IEEE 754 (Double Precision)**

1. Normalize the binary as done for 32-bit:
   - `-1.10001 × 2^3`

2. Encode:
   - Sign bit (S) = `1` (negative)
   - Exponent (E) = `3 + 1023 = 1026` → binary `10000000010`
   - Mantissa (M) = `1000100000000000000000000000000000000000000000000000`

So, the IEEE 754 64-bit representation is:

| Sign | Exponent   | Mantissa                                           |
|------|------------|----------------------------------------------------|
| 1    | 10000000010 | 1000100000000000000000000000000000000000000000000000 |


**Example 2**: Convert `0|10100100|00110100000000000000000` into decimal and otherwise

This is a 32-bit IEEE 754 representation:

1. Extract the components:
   - Sign bit (S) = `0` (positive)
   - Exponent (E) = `10100100` = `164` in decimal
   - Mantissa (M) = `1.00110100000000000000000` (1 plus the fraction after the binary point)

2. Calculate:
   - Exponent = `164 - 127 = 37`
   - Mantissa = `1 + 1/16 + 1/128 = 1.1015625`

So, the value in decimal is: $1.1015625 \times 2^37 \approx 8.517837094604492$


### To convert this back to IEEE 754 64-bit:

For the IEEE 754 64-bit (double precision):
- The value is the same, but with a larger exponent (11 bits) and more precision (52 bits for mantissa).
- The final value would be encoded similarly, but with a larger exponent bias (1023 instead of 127).

# Exhaustive Enumeration

to do

# Numerical Methods of Finding Root

Calculations with real numbers present a huge problem in computer science. Just look at the first program in this notebook, where such a simple task produced an incorrect result. The situation worsens when numbers with an infinite number of decimal places are involved, such as $\frac{1}{3} = 0.\overline{3} = 0.(3) = 0.33333...$.

In [2]:
ans = (1 / 3) + (1 / 3) + (1 / 3) + (1 / 3) + (1 / 3) + (1 / 3)
print(ans)

1.9999999999999998


One might expect that 6 * (1 / 3) would yield a similar result, but nothing could be further from the truth.

In [3]:
ans = 6 * (1 / 3)
print(ans)

2.0


As seen in this case, the computer handled the situation better when we used multiplication instead of addition. Note that this is not a rule; there are many cases where summation gives better results and others where only multiplication can preserve precision. At the beginning of learning computer science and programming, you shouldn't focus too much on this, as Python handles this problem very well. Unfortunately, not all programming languages can boast this skill. This issue is so vast that a special field of study has been created to examine this phenomenon, called **numerical methods**. However, it covers such a wide range that a solid understanding of differential calculus, linear algebra, discrete mathematics, and real analysis is needed for a good grasp, so I will not be dealing with it here.

I will focus only on what will be useful at the very beginning.

### Approximation Method

To avoid errors I assume computer make some errors during calculations. We can call this value by epsilon $\epsilon$.

- $n$: real answer (for example $\pi$)
- $\text{ans}$: computer answer (for example `3.1415926535897932`)
- $\epsilon$: possible error (for example `0.01`)

We can assume that if the following statement is `True`, then $n = \text{ans}$.

$$ |n - \text{ans}| < \epsilon \implies n \approx \text{ans} \implies n = \text{ans}$$

To illustrate this, I will tackle the problem of finding the roots of a function over a given interval. I will immediately point out that there are methods to find all the roots of a function, but these go well beyond the basics of programming, so I will focus on a simpler problem where our goal is to find one root within a specific interval. Therefore, the first thing is that we must know that there is a root within this interval.

The simplest approach is to search the entire interval by moving with a step of $\epsilon$, ensuring that every point is checked.

For this purpose, I wrote a function that returns the root.

In [4]:
def approximation_method(f, a, b, epsilon=0.01):
    if epsilon < 0:
        epsilon = -epsilon
    ans = a
    while ans <= b:
        if -epsilon <= f(ans) <= epsilon:
            return ans
        ans = ans + epsilon

def f(x):
    return x ** 2 - 4

def g(x):
    return 10 * (x ** 3) + 11 * (x ** 2) - 299 * x - 30

def h(x):
    return 2 * (x ** 3) - 3 * (x ** 2) + 5 * x

print(f'Real Answer: 2 | Approximation Method: {approximation_method(f, 1, 3)}')
print(f'Real Answer: 5 | Approximation Method: {approximation_method(g, 2, 6)}')
print(f'Real Answer: 0 | Approximation Method: {approximation_method(h, -1, 1)}')

Real Answer: 2 | Approximation Method: 2.000000000000001
Real Answer: 5 | Approximation Method: 4.999999999999936
Real Answer: 0 | Approximation Method: 7.528699885739343e-16


The main problem, however, is that this method is very, very **time-consuming**. Additionally, if $\epsilon$ is set to a small sensitivity, it may happen that the root is not found. Below is the same function, but also counting the number of iterations.

In [5]:
def approximation_method(f, a, b, epsilon=0.01):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    ans = a
    while ans <= b:
        if -epsilon <= f(ans) <= epsilon:
            return (ans, f'({iterations} iterations)')
        ans = ans + epsilon
        iterations = iterations + 1

def f(x):
    return x ** 2 - 4

def g(x):
    return 10 * (x ** 3) + 11 * (x ** 2) - 299 * x - 30

def h(x):
    return 2 * (x ** 3) - 3 * (x ** 2) + 5 * x

print(f'Real Answer: 2 | Approximation Method: {approximation_method(f, 1, 3)}')
print(f'Real Answer: 5 | Approximation Method: {approximation_method(g, 2, 6)}')
print(f'Real Answer: 0 | Approximation Method: {approximation_method(h, -1, 1)}')

Real Answer: 2 | Approximation Method: (2.000000000000001, '(100 iterations)')
Real Answer: 5 | Approximation Method: (4.999999999999936, '(300 iterations)')
Real Answer: 0 | Approximation Method: (7.528699885739343e-16, '(100 iterations)')


If we set $\epsilon = 0.000000001$, then the program's runtime and the number of iterations would increase unimaginably.

### Bisection Method

A better choice, however, is the bisection method, as it splits the search intervals in half, significantly reducing the search range with each iteration.

**Condictions**

- Function $f(x)$ must be continuous on $[a, b]$.

- Endpoint values $f(a), f(b)$ must have different signs.
    - If $f(a) > 0$,  then $f(b) < 0$.
    - If $f(a) < 0$,  then $f(b) > 0$.

**Algorithm**

1. Initialize $\text{ans} \gets \frac{a + b}{2}$ and $0 < \epsilon < 1$.

2. If $f(\text{ans}) = 0$, then $\text{ans}$ is an answer. 

    - Remember $0 <f(\text{ans}) < \epsilon \implies f(\text{ans}) = 0$.

3. If $f(\text{ans}) > 0$, then
    - if $f(a) > 0$ set $a \gets \text{ans}$,
    - otherwise set $b \gets \text{ans}$.

4. If $f(\text{ans}) < 0$, then
    - if $f(a) > 0$ set $b \gets \text{ans}$,
    - otherwise set $a \gets \text{ans}$.

5. Repeat this steps until you get the answer.

In [6]:
def bisection_method(f, a, b, epsilon=0.0001):
    if epsilon < 0:
        epsilon = -epsilon
    if a > b:
        temp = a
        a = b
        b = temp
    ans = (a + b) / 2
    while f(ans) > epsilon or f(ans) < -epsilon:
        if f(ans) > 0:
            if f(a) > 0:
                a = ans
            else:
                b = ans
        if f(ans) < 0:
            if f(a) > 0:
                b = ans
            else:
                a = ans
        ans = (a + b) / 2
    return ans

def f(x):
    return x ** 2 - 4

def g(x):
    return 10 * (x ** 3) + 11 * (x ** 2) - 299 * x - 30

def h(x):
    return 2 * (x ** 3) - 3 * (x ** 2) + 5 * x

print(f'Real Answer: 2 | Bisection Method: {bisection_method(f, 1, 3)}')
print(f'Real Answer: 5 | Bisection Method: {bisection_method(g, 2, 6)}')
print(f'Real Answer: 0 | Bisection Method: {bisection_method(h, -1, 1)}')

Real Answer: 2 | Bisection Method: 2.0
Real Answer: 5 | Bisection Method: 5.0
Real Answer: 0 | Bisection Method: 0.0


The downside of this solution, however, is that for the algorithm to work correctly, we must assume that the function values at the endpoints of the examined interval have opposite signs.

In [7]:
def bisection_method(f, a, b, epsilon=0.0001):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    if a > b:
        temp = a
        a = b
        b = temp
    ans = (a + b) / 2
    while f(ans) > epsilon or f(ans) < -epsilon:
        if f(ans) > 0:
            if f(a) > 0:
                a = ans
            else:
                b = ans
        if f(ans) < 0:
            if f(a) > 0:
                b = ans
            else:
                a = ans
        ans = (a + b) / 2
        iterations = iterations + 1
    return (ans, f'({iterations} iterations)')

def f(x):
    return x ** 2 - 4

def g(x):
    return 10 * (x ** 3) + 11 * (x ** 2) - 299 * x - 30

def h(x):
    return 2 * (x ** 3) - 3 * (x ** 2) + 5 * x

print(f'Real Answer: 2 | Bisection Method: {bisection_method(f, 1, 3)}')
print(f'Real Answer: 5 | Bisection Method: {bisection_method(g, 2, 6)}')
print(f'Real Answer: 0 | Bisection Method: {bisection_method(h, -1, 1)}')

Real Answer: 2 | Bisection Method: (2.0, '(0 iterations)')
Real Answer: 5 | Bisection Method: (5.0, '(1 iterations)')
Real Answer: 0 | Bisection Method: (0.0, '(0 iterations)')


### Newton-Raphson Method

The best method is the Newton-Raphson method because it doesn't require endpoint values, and the only requirement is an input for which the function is defined. It uses the tangent to the graph by utilizing the derivative. It determines the zero of the tangent and checks if it is also the zero of the function. A characteristic of this solution is that with each iteration, the new zero of the tangent is a better approximation of the zero of the function.

**Algorithm**

1. Set any $\text{ans}$ that $f(\text{ans})$ exists.

2. If $f(\text{ans}) = 0$, then $\text{ans}$ is an answer.

    - Remember $0 <f(\text{ans}) < \epsilon \implies f(\text{ans}) = 0$.

3. If $f(\text{ans}) \neq 0$, then $\text{ans} \gets \text{ans} - \frac{f(\text{ans})}{\frac{d}{dx}\big[ f(\text{ans}) \big]}$

4. Repeat this steps (from point 2.) until you get the answer.

In [8]:
def derivative(f, x, h=0.0001):
        return (f(x + h) - f(x)) / h

def newton_raphson_method(f, x_0, epsilon=0.0001):
    if epsilon < 0:
        epsilon = -epsilon
    ans = x_0
    while f(ans) > epsilon or f(ans) < -epsilon:
        ans = ans - (f(ans) / derivative(f, ans))
    return ans

def f(x):
    return (x ** 3) - 2 * (x ** 2) - 3 * x

print(f'Real Answer: 3 | Newton-Raphson Method: {newton_raphson_method(f, 4)}')

Real Answer: 3 | Newton-Raphson Method: 3.000000764736198


Using this method significantly reduces the number of iterations, making it a very efficient method for finding roots.

In [9]:
def derivative(f, x, h=0.0001):
        return (f(x + h) - f(x)) / h

def newton_raphson_method(f, x_0, epsilon=0.0001):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    ans = x_0
    while f(ans) > epsilon or f(ans) < -epsilon:
        ans = ans - (f(ans) / derivative(f, ans))
        iterations = iterations + 1
    return (ans, f'({iterations} iterations)')

def f(x):
    return (x ** 3) - 2 * (x ** 2) - 3 * x

print(f'Real Answer: 3 | Newton-Raphson Method: {newton_raphson_method(f, 4)}')

Real Answer: 3 | Newton-Raphson Method: (3.000000764736198, '(4 iterations)')


The problem with the Newton-Raphson method is that if during the calculations $\frac{d}{dx}\big[ f(\text{ans}) \big] = 0$, the program will return an error because division by zero is not possible. Moreover, it can be observed that in the above example, the function has more than one root, yet the function returned only one.

For example $f(x) = x^2 - 2x$ and $\text{ans} = 1$.

Iteration 1: $\text{ans} \gets 1 - \frac{-1}{0}$

In [10]:
def derivative(f, x, h=0.0001):
        return (f(x + h) - f(x)) / h

def newton_raphson_method(f, x_0, epsilon=0.0001):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    ans = x_0
    while f(ans) > epsilon or f(ans) < -epsilon:
        ans = ans - (f(ans) / derivative(f, ans))
        iterations = iterations + 1
    return (ans, f'({iterations} iterations)')

def f(x):
    return (x ** 2) - 2 * x

print(newton_raphson_method(f, 1))

(2.0000042184125744, '(17 iterations)')


In this case, the inaccuracy of the calculations helped avoid the problem, because most likely the derivative was not exactly zero, but a number very close to zero. However, this should not be treated as a rule.

### Comparision of All Numerical Methods

I will compare the number of iterations for all methods for computing the square root of $49$.

$ x = \sqrt{49}, \text{ notice that } x \geq 0$

$x = \sqrt{49} \bigg|^2 $

$x^2 = 49$

$x^2 - 49 = 0 $

My aim is to find the root of $f(x) = x^2 - 49$.

In [11]:
def approximation_method(f, a, b, epsilon=0.01):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    ans = a
    while ans <= b:
        if -epsilon <= f(ans) <= epsilon:
            return (ans, f'({iterations} iterations)')
        ans = ans + epsilon
        iterations = iterations + 1

def bisection_method(f, a, b, epsilon=0.0001):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    if a > b:
        temp = a
        a = b
        b = temp
    ans = (a + b) / 2
    while f(ans) > epsilon or f(ans) < -epsilon:
        if f(ans) > 0:
            if f(a) > 0:
                a = ans
            else:
                b = ans
        if f(ans) < 0:
            if f(a) > 0:
                b = ans
            else:
                a = ans
        ans = (a + b) / 2
        iterations = iterations + 1
    return (ans, f'({iterations} iterations)')

def derivative(f, x, h=0.0001):
        return (f(x + h) - f(x)) / h

def newton_raphson_method(f, x_0, epsilon=0.0001):
    iterations = 0
    if epsilon < 0:
        epsilon = -epsilon
    ans = x_0
    while f(ans) > epsilon or f(ans) < -epsilon:
        ans = ans - (f(ans) / derivative(f, ans))
        iterations = iterations + 1
    return (ans, f'({iterations} iterations)')

def f(x):
    return (x ** 2) - 49

print(f'Approximation Method: {approximation_method(f, 0, 10)}')
print(f'Bisection Method: {bisection_method(f, 0, 10)}')
print(f'Newton-Raphson Method: {newton_raphson_method(f, 1)}')

Approximation Method: (6.999999999999895, '(700 iterations)')
Bisection Method: (6.999998092651367, '(19 iterations)')
Newton-Raphson Method: (7.000000151432483, '(6 iterations)')


| Method                    | Description                                                                                               | Pros                                        | Cons                                           |
|---------------------------|-----------------------------------------------------------------------------------------------------------|---------------------------------------------|------------------------------------------------|
| **Approximation Method**   | This method increments the value of the starting point by a small step (`epsilon`) until the function is close to zero. | Simple to implement, works for continuous functions | Slow, may miss roots, especially if function values are close to zero. |
| **Bisection Method**       | This method divides the interval in half, narrowing down the search space based on the function's sign at the endpoints. | Guaranteed to converge (if the function is continuous and the root is within the interval). | Requires a known interval with opposite signs at the endpoints. |
| **Newton-Raphson Method**  | This method iteratively refines the root estimate using the function’s derivative to find the tangent's zero. | Fast convergence if the initial guess is close to the root. | Can fail if the derivative is zero or the initial guess is far from the root. |

